### ISSUE TITLE
求构造自定义数据集的代码和tencent数据集

### ISSUE THREAD

**Issue by ysinwell** [Kaike-Zhang/DyTed#1]
代码中只给出了对应模型输入格式的UCI数据集，能否提供将自定义的数据集（如bitcoin-alpha）转化为模型输入格式的构造代码

paper中实验的tencent数据集，我在网上没能找到，请问能否提供一下

万分感谢

**Comment by Kaike-Zhang** [comment]
---

Thank you for your inquiry.

1. **Custom Dataset Formatting**:
Given the heterogeneity of initial dataset formats, construction codes tend to differ from one dataset to another. However, the fundamental logic remains consistent: if you have fine-grained timestamps, you'd typically segment them; for coarser-grained timestamps, you might either directly employ them or opt for sampling. For your convenience, I'm providing a code snippet that deals with the Bitcoin dataset as an example:

```python
import pandas as pd
import numpy as np

import networkx as nx
import pickle as pkl

file = pd.read_csv("./data/bitcoin/soc-sign-bitcoinalpha.csv")
a = file.values.data.tolist()

time_set = set()
time = []

for b in a:
    time_set.add(b[-1])
    time.append(b[-1])

interval = (np.max(time) - np.min(time)) / len(time_set)

snap_shots = 26
interval *= (len(time_set) / snap_shots)
interval = int(interval)


def hash_int(t):
    a = t - np.min(time)
    a = int(a / interval)
    if a > snap_shots - 7:
        return snap_shots - 7
    else:
        return a

class Pair:
    def __init__(self, src, tar, t):
        self.src = src
        self.tar = tar
        self.t = t

    def get_info(self):
        return self.src, self.tar, self.t

node_set = set()
node_id = {}
id_cnt = 0

graph_edge_list = [[]] * (snap_shots - 6)
graph_node_list = [set()] * (snap_shots - 6)

Pair_list = [[]] * (snap_shots - 6)

for b in a:
    src = b[0]
    tar = b[1]
    t = hash_int(b[-1])
    p = Pair(src, tar, t)
    Pair_list[t].append(p)

graph_node_set = set()
graph = []

eg_cnt = 0

for i in range(len(Pair_list)):
    for j in range((len(Pair_list[i]))):
        src, tar, _ = Pair_list[i][j].get_info()
        if src not in node_set:
            node_set.add(src)
            node_id[src] = id_cnt
            id_cnt += 1
        if tar not in node_set:
            node_set.add(tar)
            node_id[tar] = id_cnt
            id_cnt += 1
        graph_node_set.add(node_id[src])
        graph_node_set.add(node_id[tar])
        graph_edge_list[i].append([node_id[src], node_id[tar]])
        eg_cnt += 1
    mt_Graph = nx.MultiGraph()
    temp_node_list = sorted(list(graph_node_set))
    mt_Graph.add_nodes_from(temp_node_list)
    mt_Graph.add_edges_from(graph_edge_list[i])
    graph.append(mt_Graph)
    print(len(mt_Graph.nodes))
    print(len(mt_Graph.edges))

print(len(graph))
print(len(graph_node_set))

print(eg_cnt)

with open("./data/bitcoin/graph.pkl", 'wb') as f:
    pkl.dump(graph, f)

```

Please note that this code is tailored for the Bitcoin dataset, but with slight modifications, it can cater to other datasets. 

2. **Tencent Dataset**:
As for the Tencent dataset highlighted in our paper, it contains sensitive user information. We are currently liaising with Tencent to address privacy concerns. Once the negotiations and necessary data processing are complete, we plan to make the dataset accessible to the public.

Thank you for your patience and understanding.

---

**Comment by ysinwell** [comment]
Thanks a lot

**Comment by ysinwell** [comment]
There seems to be a bug in your processing of the dataset.
In your code 
```
Pair_list = [[]] * (snap_shots - 6)

for b in a:
    src = b[0]
    tar = b[1]
    t = hash_int(b[-1])
    p = Pair(src, tar, t)
    Pair_list[t].append(p)
```

You seem to divide all edges to different snapshot.
But in fact, you copy all edges to each snapshot. 
```
>>>len(a)
24185
>>>sum([len(x) for x in Pair_list])
483700
```
483700 = edge_num * snapshot_num = 24185*20.

What's more,there are only 24185 edges in bitcoinalpha, but in your paper, the number is 483,700.I think  you truly make this faculty.

**Comment by Kaike-Zhang** [comment]
Thank you for bringing this to our attention.

I would like to clarify that among the five datasets we worked with, the processing codes were implemented by different contributors. While this allowed for diverse methodologies, it also introduced the oversight you pointed out in the Bitcoin dataset due to nuances with Python's deep and shallow copying.

Specifically, for the UCI dataset, we directly utilized the data provided by DySAT, ensuring there were no processing errors on our end. As for the remaining three datasets, our code implementation was:
```edge_date_list = [[] for _ in range(len(date_dict))]```
This method is immune to the deep or shallow copy issues.

I appreciate your keen observation, and we'll work on correcting the Bitcoin dataset in the paper.
